src/node/miner cleanups, follow-ups for #26695

stickies-v opened the pull request:
Two follow-ups for #26695, both refactoring and no observed (*) behaviour change:
- Rename `gArgs` to `args` because it's not actually a global
- Add `BlockAssembler::Options` as a (private) member to `BlockAssembler` to avoid having to assign all the options individually, essentially duplicating them

Reduces LoC and makes the code more readable, in my opinion. 

---

(*) as [pointed out by ajtowns](https://github.com/bitcoin/bitcoin/pull/26883#discussion_r1068247937), this PR changes the interface of `ApplyArgsManOptions()`, making this not a pure refactoring PR. In practice, `ApplyArgsManOptions()` is never called in such a way that this leads to observed behaviour change. Regardless, I've carved out the potential behaviour change into a separate commit and would be okay with dropping it, should it turn out to be controversial.

stickies-v committed:
refactor: rename local gArgs to args

Avoid confusion with the global gArgs

stickies-v committed:
refactor: avoid duplicating BlockAssembler::Options members

Add Options as a member to BlockAssembler to avoid having to assign
all the options individually.

Additionally brings the struct more in line with how we typically
define default and ArgManager values, as e.g. with
ChainstateManager::Options and and CTxMemPool::Options

glozow reviewed (commented):
thanks for the followup, light review ACK c14cf65ecde432c15c1e5334cfa4e5e6bdbeec92

ajtowns commented on src/node/miner.h:
```diff
@@ -172,6 +167,8 @@ class BlockAssembler
     inline static std::optional<int64_t> m_last_block_weight{};
 
 private:
+    Options m_options;
```
could make that `const` by also adding:

```c++
static BlockAssembler::Options clamp_options(const BlockAssembler::Options& opt)
{
    Options clamped{opt};
    clamped.nBlockMaxWeight = std::clamp<size_t>(clamped.nBlockMaxWeight, 4000, DEFAULT_BLOCK_MAX_WEIGHT);
    return clamped;
}
```

and setting `m_options{clamped(options)}` in the constructor.

ajtowns commented on src/node/miner.cpp:
```diff
@@ -56,34 +56,22 @@ void RegenerateCommitments(CBlock& block, ChainstateManager& chainman)
-        options.blockMinFeeRate = CFeeRate{parsed.value_or(DEFAULT_BLOCK_MIN_TX_FEE)};
-    } else {
-        options.blockMinFeeRate = CFeeRate{DEFAULT_BLOCK_MIN_TX_FEE};
+    options.nBlockMaxWeight = args.GetIntArg("-blockmaxweight", options.nBlockMaxWeight);
```
This (and below) is a potential behaviour change: if you call `ApplyArgsManOptions`, then change the arguments, and call `ApplyArgsManOptions` again on the same `options` object it won't reset values back to the defaults any more. Probably no big deal, but not really a refactor.

glozow replied:
True that there is a difference. I think this is better as it matches other `ApplyArgsManOptions` around the codebase - convention seems to be to default to the value passed in and not look at default constants. For example
https://github.com/bitcoin/bitcoin/blob/7386da7a0b08cd2df8ba88dae1fab9d36424b15c/src/node/mempool_args.cpp#L29-L38
https://github.com/bitcoin/bitcoin/blob/7386da7a0b08cd2df8ba88dae1fab9d36424b15c/src/node/validation_cache_args.cpp#L20-L32

stickies-v replied:
That's a good observation, I hadn't considered that, thanks @ajtowns. I agree with @glozow that I'd prefer this behaviour (I intentionally brought it in line with existing `ApplyArgsManOptions()` functions and `Options` objects). In practice, there is no actual behaviour change in the current code as we never use `node::ApplyArgsManOptions()` in the way you describe here. However, I'm not quite sure if it would be inappropriate to label this commit and PR with `refactor: ` - what do you think?

ajtowns replied:
Doesn't it make more sense for `ApplyArgsManOptions(args, foo)` to consistently reset `foo` to the same thing you'd get if you'd just created `foo` as a new object? TBH, I think just having an `explicit Options(const ArgsManager& args);` constructor and writing `Options cache_sizes{gArgs}` would probably be a better api...

glozow replied:
>  TBH, I think just having an explicit Options(const ArgsManager& args); constructor and writing Options cache_sizes{gArgs} would probably be a better api...

I'm mostly going with the idea from this commit message (https://github.com/bitcoin/bitcoin/commit/f1941e8bfd2eecc478c7660434b1ebf6a64095a0):

> With this patch, most CTxMemPool construction are along the lines of:
>
>    MemPoolOptions mempool_opts{...default overrides...};
>    ApplyArgsManOptions(argsman, mempool_opts);
>    ...hard overrides...
>    CTxMemPool pool{mempool_opts};
>
> This "compositional" style of building options means that we can omit
> unnecessary/irrelevant steps wherever we want but also maintain full
> customizability.
>
> For example:
>
> - For users of libbitcoinkernel, where we eventually want to remove
>  ArgsManager, they simply won't call (or even know about)
>  ApplyArgsManOptions.

As in, {mempool, block assembler, cache, etc.} `::Options` struct living in kernel/ shouldn't know about the `ArgsManager` which lives in node/. Otherwise I agree a ctor using argsman makes sense.

stickies-v committed:
miner: don't re-apply default Options value if argument is unset

ApplyArgsManOptions does not need to set default values for missing
arguments, these are already defined in the BlockAssembler::Options.

This commit changes the interface of ApplyArgsManOptions(). If
ApplyArgsManOptions() is called again after a option is changed,
this option will no longer be reset to the default value.

There is no observed behaviour change due to how
ApplyArgsManOptions() is currently used, and the new interface is
consistent with e.g. ValidationCacheSizes and MemPoolLimits.

stickies-v commented:
Force pushed to address review comments (thank you @ajtowns):
- made `m_options` `const` by adding a `ClampOptions` helper function
- carved out the `ApplyArgsManOptions()` (potential/unobserved) behaviour change into a separate commit

TheCharlatan commented on src/node/miner.h:
```diff
@@ -6,6 +6,7 @@
 #ifndef BITCOIN_NODE_MINER_H
 #define BITCOIN_NODE_MINER_H
 
+#include <policy/policy.h>
```
Nit: IWYU says you can now remove this include from `miner.cpp`

sedited reviewed (commented):
Light code review ACK 6a5e88e5cf06a6b410486cc36aba7afece0d9da9

glozow commented:
ACK 6a5e88e5cf
